workload-update: implement bakoff mechanism

alicefr opened the pull request:
When the migration fails, the workload update continues to create new VMI migration object. Now, virt-controller implement the migration backoff mechanism also for workload update migrations

Add a functional test which replicates a migration failure when the destination image is smaller then source one.

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

alicefr commented:
/cc @awels @mhenriks

alicefr commented:
@awels right now if we encounter a failure in the migration, then on the VMI has the VolumeChange condition set to false with the error message  "The migration has failed". Would this be enough for you or do you want something more precise? In the VMI status there is also more detail with the reason of the failure, putting that info here as well is a little bit redundant. WDYT?

awels commented:
I am fine with just the migration has failed message. I can look up the additional information if I need it on my end.

awels commented on tests/storage/migration.go:
```diff
@@ -365,6 +369,20 @@ var _ = SIGDescribe("[Serial]Volumes update with migration", Serial, func() {
 			waitMigrationToNotExist(vm.Name, ns)
 		})
+
+		It("should failed to migrate when the destination image is smaller", func() {
```
```suggestion
		It("should fail to migrate when the destination image is smaller", func() {
```

awels commented on tests/storage/migration.go:
```diff
@@ -365,6 +369,20 @@ var _ = SIGDescribe("[Serial]Volumes update with migration", Serial, func() {
+			Eventually(func() bool {
+				vmi, err := virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				return controller.NewVirtualMachineInstanceConditionManager().HasConditionWithStatus(vmi,
```
Maybe return the message here and check if it is `The migration has failed`

awels reviewed (commented):
looks good some nitpicks.

fossedihelm commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -2302,13 +2305,23 @@ func (c *VMIController) requireVolumesUpdate(vmi *virtv1.VirtualMachineInstance)
+			Message:            "The migration has failed",
+		})
+	} else {
+
```
nit: newline

fossedihelm commented on tests/storage/migration.go:
```diff
@@ -365,6 +369,20 @@ var _ = SIGDescribe("[Serial]Volumes update with migration", Serial, func() {
 		})
+
+		It("should failed to migrate when the destination image is smaller", func() {
+			volName := "disk0"
```
`const`

fossedihelm commented on tests/storage/migration.go:
```diff
@@ -377,3 +395,73 @@ func createUnschedulablePVC(name, namespace, size string) *k8sv1.PersistentVolum
+// createSmallImageForDestinationMigration creates a smaller raw image on the destination PVC and the PVC is bound to another node then the running
+// virt-launcher in order to allow the migration.
+func createSmallImageForDestinationMigration(vm *virtv1.VirtualMachine, name, size string) *k8sv1.PersistentVolumeClaim {
+	volName := "vol"
+	dir := "disks"
```
`const`

fossedihelm commented on tests/storage/migration.go:
```diff
@@ -377,3 +395,73 @@ func createUnschedulablePVC(name, namespace, size string) *k8sv1.PersistentVolum
+
+// createSmallImageForDestinationMigration creates a smaller raw image on the destination PVC and the PVC is bound to another node then the running
+// virt-launcher in order to allow the migration.
+func createSmallImageForDestinationMigration(vm *virtv1.VirtualMachine, name, size string) *k8sv1.PersistentVolumeClaim {
```
The returned pvc is not used, you can drop it

fossedihelm reviewed (commented):
Thank you @alicefr!
Some comments below

alicefr commented on tests/storage/migration.go:
```diff
@@ -377,3 +395,73 @@ func createUnschedulablePVC(name, namespace, size string) *k8sv1.PersistentVolum
+		p, err = virtCli.CoreV1().Pods(vmi.Namespace).Get(context.Background(), p.Name, metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		return p.Status.Phase == k8sv1.PodSucceeded
+	}, 120*time.Second, time.Second).Should(BeTrue())
```
Done

fossedihelm commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -47,6 +48,8 @@ import (
 	"k8s.io/client-go/util/workqueue"
 	"k8s.io/utils/trace"
 
+	"kubevirt.io/kubevirt/pkg/util/migrations"
```
Sorry for noticing this late:
Can you move the import to the proper section? In the `kubevirt.io/kubevirt/...` block below

fossedihelm reviewed (commented):
Last nit and we are ready to go from my point of view

fossedihelm reviewed (approved):
Thank you for your patience

mhenriks commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -2302,13 +2304,22 @@ func (c *VMIController) requireVolumesUpdate(vmi *virtv1.VirtualMachineInstance)
+			LastTransitionTime: v1.Now(),
+			Status:             k8sv1.ConditionTrue,
+			Message:            "migrate volumes",
+		})
```
Say a user fixes whatever got the migration to fail and wants it to run again.  What do they do?

mhenriks replied:
I understand that current behavior may not be great but I think an exponential backoff strategy would be better than one and done

alicefr replied:
I'm not sure, I think if the migration fails you cannot repair it

mhenriks replied:
Why because of data on the target pvc? When a user retries migration do they also have to make sure to clean the target pvc?

alicefr replied:
No, because the size of the destination could be wrong. In the functional test, I have created the image manually, but we can imagine that this might happen if the destination storage class has a larger overhead. Without this fix, the workload updater continuously to create migration objects and new pods using the destination pvc. I find it hard to be able to fix the storage while the update is in progress.

Additionally, it really pollutes the systems by creating pods in an infinite loop.

mhenriks replied:
> No, because the size of the destination could be wrong. 

Sure there are cases where the user configured things wrong, but what about transient errors where retrying would fix things?

> Additionally, it really pollutes the systems by creating pods in an infinite loop.

Of course which is why I am advocating for exponential retry.  That is the kubernetes way.  Transient errors will eventually converge.  If the user creates a bad config then retries will eventually happen infrequently enough to no be a big burden

alicefr replied:
Ah, I see. I need to figure out how/where to save the number of retries

mhenriks replied:
Can you simply take advantage of built in exponential backoff when error is returned in controller reconcile loop?

alicefr replied:
Not sure how.
The code here is actually wrong and it isn't the right place the vmi controller. I'm developing a new version moving this in the workload updater controller.

alicefr commented:
@awels @mhenriks @fossedihelm could you please take another look to this. I have changed the implementation. Now, we have a new field in the VMI status which keeps track of the number of retries. I also added the exponential retries suggested by Mike in https://github.com/kubevirt/kubevirt/pull/12336#discussion_r1681145118

fossedihelm commented:
@alicefr We already have a backoff mechanism for evacuation migrations.
How about adding an annotation for volume update migrations and extend the behavior of `handleMigrationBackoff` ?https://github.com/kubevirt/kubevirt/blob/69c1833597ee13f199bcd8aee098f06c853b6ce3/pkg/virt-controller/watch/migration.go#L1267

This would save us a lot of ad hoc code. Have you considered that? Thank you

alicefr commented:
> @alicefr We already have a backoff mechanism for evacuation migrations. How about adding an annotation for volume update migrations and extend the behavior of `handleMigrationBackoff` ?
> 
> https://github.com/kubevirt/kubevirt/blob/69c1833597ee13f199bcd8aee098f06c853b6ce3/pkg/virt-controller/watch/migration.go#L1267
> 
> This would save us a lot of ad hoc code, and of course an API change. Have you considered that? Thank you

I didn't know it, I'll look into it, thanks for the suggestion

alicefr commented:
@fossedihelm would you mind taking another look? It now uses the backoff mechanism

fossedihelm commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -1742,6 +1749,13 @@ func (c *MigrationController) listEvacuationMigrations(namespace string, name st
 	})
 }
 
+func (c *MigrationController) listWorkloadUpdaterMigrations(namespace string, name string) ([]*virtv1.VirtualMachineInstanceMigration, error) {
+	return c.filterMigrations(namespace, func(migration *virtv1.VirtualMachineInstanceMigration) bool {
+		_, isWorkloadUpdater := migration.Annotations[virtv1.WorkloadUpdateMigrationAnnotation]
+		return migration.Spec.VMIName == name && isWorkloadUpdater
+	})
+}
+
```
How about reducing loop execution?
```

func (c *MigrationController) listBackoffEligibleMigrations(namespace string, name string) ([]*virtv1.VirtualMachineInstanceMigration, error) {
	return c.filterMigrations(namespace, func(migration *virtv1.VirtualMachineInstanceMigration) bool {
		return evacuationMigrationsFilter(migration, name) && workloadUpdaterMigrationsFilter(migration, name)
	})
}

func evacuationMigrationsFilter(migration *virtv1.VirtualMachineInstanceMigration, name string) bool {
	_, isEvacuation := migration.Annotations[virtv1.EvacuationMigrationAnnotation]
	return migration.Spec.VMIName == name && isEvacuation
}

func workloadUpdaterMigrationsFilter(migration *virtv1.VirtualMachineInstanceMigration, name string) bool {
	_, isWorkloadUpdater := migration.Annotations[virtv1.WorkloadUpdateMigrationAnnotation]
	return migration.Spec.VMIName == name && isWorkloadUpdater
}
```

fossedihelm reviewed (commented):
A little performance improvement suggestion.
Also, can we add a unit test under the `Migration backoff` Context?
Thank you!

alicefr commented:
> A little performance improvement suggestion. Also, can we add a unit test under the `Migration backoff` Context? Thank you!

@fossedihelm applied the last comments. PTAL

fossedihelm commented:
@alicefr Thank you!
/lgtm
Can you adjust the PR title :)

alaypatel07 commented:
/remove-label kind/api-change

mhenriks commented:
/approve

Great that you could get this to work with existing backoff machinery

alicefr commented:
/cherrypick release-1.3
